Feature/units

petersteneteg opened the pull request:
Thank you for contributing to the Inviwo repository! Great code makes great apps so please ensure the following:

**Always**
- [ ] Code is error and warning free
- [ ] Code follows the [Inviwo coding guidelines](https://github.com/inviwo/inviwo/wiki/Coding-Conventions)

**New features**
- [ ] Code / Processors follows the [Inviwo documentation guidelines](https://github.com/inviwo/inviwo/wiki/Documentation-guide)
- [ ] Code / processors are tested in [unittest and/or regresion tests](https://github.com/inviwo/inviwo/wiki/Manual#testing)

**Updated feature / Fixes** 
- [ ] Relevant documentation has been updated

**What evineroment has the code been compiled and tested on?** 
- Operating system
- Compiler/IDE
- QT/CMake/Python versions

Are you are unsure how to address these points? Please make a note here so that we can help out:

petersteneteg committed:
Core: Added some support for specifying units and propagating them. Uses the LLNL/Units lib to parse and do runtime unit operations.

Also use string_view in most property constructors.

petersteneteg committed:
Core/DataFrame/Plotting: More work on integrating units.

martinfalk committed:
Core: DataMapper documentation and  Format fixes

martinfalk commented on include/inviwo/core/algorithm/permutations.h:
```diff
@@ -0,0 +1,226 @@
+namespace util {
+
+template <typename T>
+std::vector<T> makeVectorRange(ptrdiff_t start, ptrdiff_t end, ptrdiff_t step = 1) {
```
great!

martinfalk commented on include/inviwo/core/algorithm/permutations.h:
```diff
@@ -0,0 +1,226 @@
+ *
+ * Example:
+ * \code
+ * std::array values = {0, 1, 2};
```
missing type and size for std::array

petersteneteg replied:
no it gets deduced

martinfalk commented on include/inviwo/core/algorithm/permutations.h:
```diff
@@ -0,0 +1,226 @@
+ *
+ * Example:
+ * \code
+ * std::array values = {0, 1, 2};
```
missing type and size for std::array

petersteneteg replied:
no it gets deduced

martinfalk commented on include/inviwo/core/algorithm/permutations.h:
```diff
@@ -0,0 +1,226 @@
+ *
+ * Example:
+ * \code
+ * std::array sizes = {3, 2};
```
missing type and size for std::array

petersteneteg replied:
no it gets deduced

martinfalk commented on include/inviwo/core/datastructures/unitsystem.h:
```diff
@@ -0,0 +1,344 @@
+ *
+ * format_spec ::= fmt standard format specifier
+ *
+ * unis_spec   ::= [prefix][units]
```
`unit_spec`?

martinfalk commented on include/inviwo/core/datastructures/unitsystem.h:
```diff
@@ -0,0 +1,344 @@
+
+    // All the unicode superscript digits from 0 to 9 but we let 0,1 ("\u2070", "\u00B9") be empty
+    // since they are not needed
+    static constexpr std::array<std::string_view, 10> powers = {
```
what about numbers beyond 9? Don't they need "0" and "1"?

petersteneteg replied:
0 is never used m^1 = m and Ive never seen m^10 also the unitlib does not handle powers that large anyway...

martinfalk commented on include/inviwo/core/util/stringconversion.h:
```diff
@@ -269,6 +270,13 @@ constexpr std::string_view trim(std::string_view str) noexcept {
 
+constexpr auto fmtHelp = IVW_UNINDENT(R"(
+    To locate the error try breaking on fmt::format_error exceptions,
+    or put a breakpoint in fmt::detail::error_handler::on_error.
```
That is some very specific information.

petersteneteg replied:
Yes, but it is hard to figure out what is wrong when you get a message "invalid format" and it can be any format string in all of Inviwo that is to blame... Might at the relevant MVSC and GCC commands to at some point...

martinfalk commented on modules/base/include/modules/base/algorithm/volume/volumelaplacian.h:
```diff
@@ -143,7 +145,10 @@ std::shared_ptr<Volume> VolumeLaplacianDispatcher::operator()(
+    newVolume->axes = volume->axes;
+    newVolume->dataMap_.valueAxis.name = "laplacian";
+    newVolume->dataMap_.valueAxis.unit =
+        volume->dataMap_.valueAxis.unit / volume->axes[0].unit / volume->axes[0].unit;
```
is this correct? Shouldn't there be three units below the fraction?

martinfalk replied:
What about the different axes? Are you assuming they are identical?

petersteneteg replied:
no it is a second derivative with respect to space, I assume all dimensions have the same unit other wise it makes no sense...

martinfalk commented on modules/base/src/io/datvolumesequencereader.cpp:
```diff
@@ -188,29 +193,69 @@ std::shared_ptr<DatVolumeSequenceReader::VolumeSequence> DatVolumeSequenceReader
+        {"axis2name", [](State& state, std::stringstream& ss) { ss >> state.axes[1].name; }},
+        {"axis3name", [](State& state, std::stringstream& ss) { ss >> state.axes[2].name; }},
+        
+        {"axesunits",
```
are these documented somewhere for the "dat" files?

petersteneteg replied:
I think you are looking at the documentation. PR would be welcome. I just made this up, I've never seen any "official" dat files

martinfalk replied:
That's what I would have assumed.

martinfalk commented on modules/base/src/io/datvolumewriter.cpp:
```diff
@@ -127,7 +138,7 @@ void writeDatVolume(const Volume& data, const std::string filePath, bool overwri
                                   IVW_CONTEXT_CUSTOM("util::writeDatVolume"));
     }
-}
+}  // namespace inviwo
```
revert those two lines. This is not closing a namespace.

martinfalk commented on modules/dataframe/include/inviwo/dataframe/datastructures/column.h:
```diff
@@ -129,18 +132,25 @@ class TemplateColumn : public Column {
-                   std::shared_ptr<Buffer<T>> buffer = std::make_shared<Buffer<T>>());
+    explicit TemplateColumn(std::string_view header,
+                            std::shared_ptr<Buffer<T>> buffer = std::make_shared<Buffer<T>>(),
+                            Unit unit = Unit{}, std::optional<dvec2> range = std::nullopt);
```
neat!

martinfalk commented on modules/dataframe/include/inviwo/dataframe/datastructures/column.h:
```diff
@@ -129,18 +132,25 @@ class TemplateColumn : public Column {
+                   std::optional<dvec2> range = std::nullopt);
 
     TemplateColumn(const TemplateColumn<T>& rhs);
+    TemplateColumn(const TemplateColumn<T>& rhs, const std::vector<size_t>& rowSelection);
```
`std::vector<size_t>` might clash with the selection returned from a BitSet (`std::vector<uint32_t>`).

petersteneteg replied:
ahh might be better to use unit32_t... will change that

martinfalk commented on modules/dataframe/include/inviwo/dataframe/datastructures/dataframe.h:
```diff
@@ -255,13 +266,14 @@ struct DataTraits<DataFrame> {
+            if (col->getCustomRange()) {
+                return toString(*col->getCustomRange());
             } else {
                 return std::string("-");
```
Hm, should we instead just always use `getRange()` here? The main reason for the if statement was the optional range.

petersteneteg replied:
think we also print the min max though

martinfalk commented on modules/dataframe/include/inviwo/dataframe/datastructures/dataframe.h:
```diff
@@ -288,64 +300,22 @@ struct DataTraits<DataFrame> {
+                    if (col_q->getSize() == 0) return {"-", "-"};
+
+                    const auto minmax = col_q->getDataRange();
+                    return {fmt::format("{:8.4g}", minmax.x), fmt::format("{:8.4g}", minmax.y)};
```
is this equivalent to std::defaultfloat?

petersteneteg replied:
what does std::defaultfloat? even do? seems like some reset?

martinfalk replied:
Sets the floatfield of the stream str to zero, as if by calling str.unsetf(std::ios_base::floatfield). This enables the default floating-point formatting, which is different from fixed and scientific.

See https://en.cppreference.com/w/cpp/io/manip/fixed

```
┌──────────┬────────────┬──────────────────────────┐
│  number  │   iomanip  │      representation      │
├──────────┼────────────┼──────────────────────────┤
│ 0.0      │ fixed      │ 0.000000                 │
│ 0.0      │ scientific │ 0.000000e+00             │
│ 0.0      │ hexfloat   │ 0x0p+0                   │
│ 0.0      │ default    │ 0                        │
├──────────┼────────────┼──────────────────────────┤
│ 0.01     │ fixed      │ 0.010000                 │
│ 0.01     │ scientific │ 1.000000e-02             │
│ 0.01     │ hexfloat   │ 0x1.47ae147ae147bp-7     │
│ 0.01     │ default    │ 0.01                     │
├──────────┼────────────┼──────────────────────────┤
│ 0.00001  │ fixed      │ 0.000010                 │
│ 0.00001  │ scientific │ 1.000000e-05             │
│ 0.00001  │ hexfloat   │ 0x1.4f8b588e368f1p-17    │
│ 0.00001  │ default    │ 1e-05                    │
└──────────┴────────────┴──────────────────────────┘
```

petersteneteg replied:
Same as g I think

martinfalk commented on modules/qtwidgets/src/tf/tfeditorprimitive.cpp:
```diff
@@ -229,10 +231,11 @@ void TFEditorPrimitive::updateLabel() {
+        label = utilqt::toQString(
+            fmt::format("{:0.3g} {} ({:0.3g}) / {:0.3g}",
+                        tfe->getDataMapper().mapFromNormalizedToValue(getPosition()),
+                        tfe->getDataMapper().valueAxis.unit, getPosition(), getColor().a));
```
nitpicking: additional space if there is no unit.

petersteneteg replied:
have a fix on the way

martinfalk commented on src/core/tests/unittests/unitsystem-test.cpp:
```diff
@@ -0,0 +1,62 @@
+    std::cout << fmt::format("{}", chargeDensity) << "\n";
+    std::cout << fmt::format("{}", chargeDensity * length.pow(3)) << "\n";
+
+    std::cout << fmt::format("{0:all:<20}", chargeDensity) << "\n";
```
no check on result?

petersteneteg replied:
WIP

petersteneteg committed:
Core/Various: More unit work, review comments

petersteneteg committed:
Core: Unit system tests

ResearchDaniel commented on include/inviwo/core/algorithm/linearmap.h:
```diff
@@ -0,0 +1,67 @@
+
+template <typename T>
+constexpr T linearMap(T val, glm::vec<2, T> from, glm::vec<2, T> to) noexcept {
+    return (val - from.x) / (from.y - from.x) * (to.y - to.x) + to.x;
```
I think that these functions will only work for floating point numbers. They either should be specialized for other types (and cast the denominator to floating point number) or only allowed for floating point numbers

ResearchDaniel commented on include/inviwo/core/algorithm/linearmap.h:
```diff
@@ -0,0 +1,67 @@
+template <typename T, size_t N>
+constexpr glm::vec<N, T> linearMap(glm::vec<N, T> val, glm::vec<2, T> from,
+                                   glm::vec<2, T> to) noexcept {
+    return (val - from.x) / (from.y - from.x) * (to.y - to.x) + to.x;
```
See comment above

ResearchDaniel commented on include/inviwo/core/algorithm/linearmap.h:
```diff
@@ -0,0 +1,67 @@
+
+template <typename T, size_t N>
+constexpr glm::vec<N, T> linearMapToNormalized(glm::vec<N, T> val, glm::vec<2, T> from) noexcept {
+    return (val - from.x) / (from.y - from.x);
```
See comment above

ResearchDaniel commented on include/inviwo/core/algorithm/linearmap.h:
```diff
@@ -0,0 +1,67 @@
+
+template <typename T>
+constexpr T linearMapToNormalized(T val, glm::vec<2, T> from) noexcept {
+    return (val - from.x) / (from.y - from.x);
```
See comment above

ResearchDaniel commented on include/inviwo/core/algorithm/permutations.h:
```diff
@@ -0,0 +1,226 @@
+        , hasMore{true} {
+
+        if (r > pool.size() || r < 1) {
+            throw Exception("Invalid permutation size larger then input");
```
Type, then-> than

ResearchDaniel commented on include/inviwo/core/algorithm/permutations.h:
```diff
@@ -0,0 +1,226 @@
+        : pool{values}, indices{makeVectorRange<ptrdiff_t>(0, r)} {
+
+        if (indices.size() > pool.size() || indices.empty()) {
+            throw Exception("Invalid combinations size larger then input");
```
Than

ResearchDaniel commented on include/inviwo/core/datastructures/data.h:
```diff
@@ -182,6 +182,13 @@ class Data {
+ * Tag type for constructors indicating that you want a copy of only "meta" data not the actual
+ * data.
+ */
+struct NoData {};
```
OnlyMetaData would be more explicit?

ResearchDaniel replied:
After going through all changes I think that NoData is ok

ResearchDaniel commented on include/inviwo/core/datastructures/datamapper.h:
```diff
@@ -54,36 +55,87 @@ class IVW_CORE_API DataMapper {
-    dvec2 valueRange;       ///< Minimum and maximum value range
-    std::string valueUnit;  ///< Unit, i.e. Hounsfield/absorption/W.
+    /**
+     * The ´dataRange´ is used to normalize the "raw" values to [0,1]
```
You could say that it is used to map the stored (raw) values to the normalized range [0 1].
Small difference but also includes the reason for having it

petersteneteg replied:
I don't think stored has the right meaning that I want to express.

ResearchDaniel commented on include/inviwo/core/datastructures/datamapper.h:
```diff
@@ -54,36 +55,87 @@ class IVW_CORE_API DataMapper {
+    dvec2 dataRange;
+
+    /**
+     * The `valueRange` corresponds to the "true" values in the "raw" data. The "raw" data will
```
Might want to adjust: true (physical) 
I think that stored might be a better word than raw

petersteneteg replied:
physical does not make sense to me.

ResearchDaniel commented on include/inviwo/core/datastructures/unitsystem.h:
```diff
@@ -0,0 +1,379 @@
+
+#include <flags/flags.h>
+#include <fmt/format.h>
+#include <units/units.hpp>
```
Do we need to ignore warnings for these?

ResearchDaniel commented on include/inviwo/core/datastructures/unitsystem.h:
```diff
@@ -0,0 +1,379 @@
+ *    "["     Surround the unit in [unit]
+ *
+ * Prefix:
+ *    "p"     Use si prefixes y to Y to reduce any multiplier (default)
```
Do you mean SI?

ResearchDaniel commented on include/inviwo/core/datastructures/unitsystem.h:
```diff
@@ -0,0 +1,379 @@
+ *
+ * Units:
+ *    "si"    Only use the basic SI unit and combination of those.
+ *    "ext"   Use unit from the si, derived and extra grops.
```
SI

ResearchDaniel commented on include/inviwo/core/datastructures/unitsystem.h:
```diff
@@ -0,0 +1,379 @@
+struct fmt::formatter<::inviwo::Unit> {
+    formatter<std::string_view> formatter_;
+
+    enum class UnitSystem { Si, Ext, Sys, All };
```
Should we use SI instead of Si?

ResearchDaniel commented on modules/basegl/src/processors/volumeprocessing/volumegradientmagnitude.cpp:
```diff
@@ -56,7 +56,13 @@ void VolumeGradientMagnitude::preProcess(TextureUnitContainer&) {
+    volume_->dataMap_.valueAxis.name = "gradient magnitude";
+    volume_->dataMap_.valueAxis.unit =
+        inport_.getData()->dataMap_.valueAxis.unit / inport_.getData()->axes[0].unit;
+    volume_->dataMap_.dataRange = dvec2(0.0, 1.0);
```
Are the computed gradients normalized?

petersteneteg replied:
Nope I don't think so, this range is bogus, but I did not change that...

ResearchDaniel commented on modules/basegl/src/processors/volumeprocessing/volumegradientprocessor.cpp:
```diff
@@ -70,13 +70,19 @@ VolumeGradientProcessor::VolumeGradientProcessor()
+    volume_->dataMap_.valueAxis.name = "gradient";
+    volume_->dataMap_.valueAxis.unit =
+        inport_.getData()->dataMap_.valueAxis.unit / inport_.getData()->axes[0].unit;
+    volume_->dataMap_.dataRange = dvec2(-1.0, 1.0);
```
Also here I am not sure that this will be correct unless the gradients are normalized

petersteneteg replied:
Nope I don't think so, this range is bogus, but I did not change that...

ResearchDaniel commented on modules/dataframe/include/inviwo/dataframe/datastructures/column.h:
```diff
@@ -64,12 +66,39 @@ class IVW_MODULE_DATAFRAME_API Column : public MetaDataOwner {
+    virtual dvec2 getDataRange() const = 0;
+
+    /**
+     * Returns the custom range if set or else the range of the data
```
The range of the data = min/max of the data? Might want to state that instead in that case

ResearchDaniel commented on modules/dataframe/include/inviwo/dataframe/datastructures/dataframe.h:
```diff
@@ -97,25 +98,32 @@ class IVW_MODULE_DATAFRAME_API DataFrame : public MetaDataOwner {
      * \brief add column based on the contents of the given buffer
      * updateIndexBuffer() needs to be called after all columns have been added before
      * the DataFrame can be used
+     * Note: this will copy the data of buffer.
```
The why do we input a shared ptr?

petersteneteg replied:
I have absolutely no Idea

ResearchDaniel commented on include/inviwo/core/algorithm/permutations.h:
```diff
@@ -0,0 +1,226 @@
+ * \endcode
+ */
+template <typename T>
+class Permutations {
```
Do these new algorithms need to be in core?

petersteneteg replied:
Yes, I needed them for the unit system

ResearchDaniel reviewed (changes requested):
Nice! Some remarks that should be double checked

petersteneteg committed:
Core: Unit fixes and review update, Added a String_s_Property for a 2-4 strings

ResearchDaniel commented on include/inviwo/core/algorithm/linearmap.h:
```diff
@@ -35,32 +35,38 @@ namespace inviwo::util {
 
 template <typename T>
 constexpr T linearMapFromNormalized(T val, glm::vec<2, T> to) noexcept {
+    static_assert(std::is_floating_point_v<T>);
```
This one does not need to be floating point.
Also, will this not break code that utilize data dispatch functionality?
In most cases it won't make sense to perform computations using the type they are stored.
In my opinon it would probably be better to return float/double, but perhaps another PR as this does not really change the existing behaviour.

petersteneteg replied:
This is new code so it not used outside of this PR

martinfalk committed:
Core: MSVC compile fixes

martinfalk committed:
Core,DataFrame,PlottingGL,QtWidgets: MSVC compile and warning fixes

martinfalk committed:
BrushingAndLinking: port invalidation

* invalidate inport if there is no parent manager to trigger a re-evaluation of the processor

martinfalk committed:
DataFramePython: updated bindings

martinfalk committed:
DataFrameQt: TableView fix

* default sort order set to ascending

petersteneteg committed:
Core: Property setDefault(T), Lots for fixes in many places

Added a setDefautl(T state) to most Properties, to be able to change the default without triggerign a property modified call.

petersteneteg committed:
Plotting: Volume axis fix

martinfalk commented on modules/dataframe/src/processors/csvsource.cpp:
```diff
@@ -55,29 +56,57 @@ CSVSource::CSVSource(const std::string& file)
+
+        auto dataFrame = std::make_shared<DataFrame>(*loadedData_);
+        columns_.updateDataFrame(*dataFrame);
+        data_.setData(dataFrame);
```
shouldn't there be a `loadingFailed_ = false;`?

One might fix the file and trigger a processor invalidation using the reload button property or invalidate the processor directly.

petersteneteg replied:
Done in the onChange calls

martinfalk committed:
DataFrame: persistent row IDs for joins

petersteneteg committed:
Units: version update, pybindings fixes

petersteneteg committed:
Plotting/DataFame: Pixel Align text, handle csvsource loading failures

Inviwo-Team committed:
Jenkins: Format fixes

petersteneteg committed:
PlottingGL: Align text regression update

# Conflicts:
#	modules/plottinggl/tests/regression/scatterplot/Canvas.png

martinfalk commented on include/inviwo/core/datastructures/volume/volume.h:
```diff
@@ -87,15 +89,13 @@ class IVW_CORE_API Volume : public Data<Volume, VolumeRepresentation>,
      * @param format The format of the data.
      */
-    // clang-format off
-    [[ deprecated("use VolumeRepresentation::setDataFormat() instead (deprecated since 2019-06-26)")]]
```
is there a particular reason for removing the deprecation?

petersteneteg replied:
I needed to use it :)

martinfalk replied:
that's a valid reason :-)

martinfalk reviewed (dismissed):
Great work. So many new features and utility functions I am looking forward to use :-)
units, `util::ref`, cloning only certain rows of columns, ...

Don't forget to update the Changelog.

ResearchDaniel commented on data/volumes/boron.dat:
```diff
@@ -2,8 +2,12 @@ Rawfile: boron.raw
 Format: UINT16
 DataRange: 0 65535
-ValueRange: 0 65535
+ValueRange: 0.0 0.000098763573
```
Should also change testdata/boron.ivf and nice-to-have would be to also change Bi2O3.dat

ResearchDaniel commented on ext/units/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+set(UNITS_BUILD_WEB_SERVER OFF)
+set(UNITS_BUILD_FUZZ_TARGETS OFF)
+set(UNITS_ENABLE_ERROR_ON_WARNINGS OFF)
+set(UNITS_ENABLE_EXTRA_COMPILER_WARNINGS ON)
```
Do we want this to be on?

petersteneteg replied:
Did not see any reason not to leave it on, did not get any odd warnings

ResearchDaniel commented on include/inviwo/core/properties/stringsproperty.h:
```diff
@@ -0,0 +1,138 @@
+ * \brief CompositeProperty for a fixed set of StringProperties
+ */
+template <size_t N>
+class StringsProperty : public CompositeProperty {
```
Using (public/protected) inheritance here means that it will be possible to add other properties to the StringsProperty. 
I think that we have some of these situations where it should not be possible, but not sure if we currently have a solution for it.

petersteneteg replied:
not much we can do here

ResearchDaniel commented on include/inviwo/core/properties/stringsproperty.h:
```diff
@@ -0,0 +1,138 @@
+    auto b = doc.get({P("html"), P("body")});
+
+    utildoc::TableBuilder tb(b, P::end(), {{"String", "Value"}});
+    for (auto& prop : strings) {
```
```suggestion
    for (const auto& prop : strings) {
```

ResearchDaniel commented on include/inviwo/core/util/settings/unitsettings.h:
```diff
@@ -0,0 +1,45 @@
+
+namespace inviwo {
+
+class IVW_CORE_API UnitSettings : public Settings {
```
```suggestion
/**
 *
 * \brief Application-wide Settings for axis units, e.g. the SI unit meter.
 * @see unitgroups for available units.
 */
class IVW_CORE_API UnitSettings : public Settings {
```

ResearchDaniel commented on include/inviwo/core/datastructures/datamapper.h:
```diff
@@ -54,36 +55,87 @@ class IVW_CORE_API DataMapper {
+     * The `valueAxis` holds the `name` and `unit` of the quantity. Following the example above
+     * the `name` would be "water temperature" and `unit` "°C".
+     */
+    Axis valueAxis;  ///< Name and Unit, i.e. "absorption", "Hounsfield".
```
Not sure this belongs here as it adds extra responsibility to this class. Sure, it is related but also quite hidden.
It would be more visible if we put it next to the axis units of the Image/Volume as we have done for the Column. Would that make more sense?

petersteneteg replied:
We had a "unit" string in there before. Think that is the main reason for me putting it there. It is also related to the valueRange. Since the unit applies only for "raw" values mapped to the valueRange.

ResearchDaniel reviewed (dismissed):
Some comments and suggestions. 

The most important one being about where to place the value unit. It might be better of next to the axis members in my opinion

petersteneteg committed:
Plotting: Minor fixes and refactorings. Added a LabelType to the colorScaleLegend

petersteneteg committed:
Plotting: VolumeAxis got automatic axis selection

petersteneteg committed:
Qt: fixes to avoid extra evaluations while editing

Inviwo-Team committed:
Jenkins: Format fixes

petersteneteg committed:
Python: format PEP8 formatting, attempt to fix/debug web browser sync issues.

petersteneteg committed:
Python: PEP8 Format fixes, FontRendering: TextOverlay handles many labels and better arg handling

petersteneteg committed:
FontRendering: Doc update

Inviwo-Team committed:
Jenkins: Format fixes

martinfalk commented on modules/fontrendering/include/modules/fontrendering/processors/textoverlaygl.h:
```diff
@@ -65,12 +82,12 @@ class ImageGL;
  *   * __Position__ Where to put the text, relative position from 0 to 1
- *   * __Anchor__ What point of the text to put at "Position". relative from -1,1. 0 meas the image
- *     is centered on "Position".
+ *   * __Anchor__ What point of the text to put at "Position". relative from -1,1. 0 meas the
```
-> "Relative from -1,1. 0 means the"

martinfalk commented on modules/fontrendering/src/processors/textoverlaygl.cpp:
```diff
@@ -59,49 +82,37 @@ TextOverlayGL::TextOverlayGL()
     , enable_("enable", "Enabled", true)
-    , text_("text", "Text", "Lorem ipsum etc.", InvalidationLevel::InvalidOutput,
-            PropertySemantics::TextEditor)
+    , texts_{"texts", "Texts", std::make_unique<TextOverlayProperty>("text0", "Text 0"), 99,
```
why the 99 limit here?

petersteneteg replied:
you had it in there before since you only parsed %99, think I have removed it now...

martinfalk replied:
for the args, yes. But not the single text line.

petersteneteg committed:
Python: Added ListProperty Bindings, some regression fixes

Inviwo-Team committed:
Jenkins: Format fixes

petersteneteg committed:
FontRendering/Plotting: Regression fixes

petersteneteg committed:
Core: Review fixes

petersteneteg committed:
Core: removed duplicated code

petersteneteg committed:
Core: Changelog update, regression fixes

# Conflicts:
#	CHANGELOG.md
#	resources/changelog.html

petersteneteg committed:
PlottingGL: Regression update

martinfalk commented on CHANGELOG.md:
```diff
@@ -1,6 +1,15 @@
 
-## 2021-11-23 CompositeProperty collapse & expand
+## 2021-12-09 DataFrame Column Units
+The `DataFrame` `Column` now has a `unit` member and many processors (`PCP`, `ScatterPlot`, `DataFrameTable`), operating on DataFrames will now also display the column units if available. The `ColumnMetaDataProperty` used in the `CSVSourc` processor got support for specifying units. 
```
typo `CSVSourc_e_`

martinfalk commented on CHANGELOG.md:
```diff
@@ -1,6 +1,15 @@
+The `DataFrame` `Column` now has a `unit` member and many processors (`PCP`, `ScatterPlot`, `DataFrameTable`), operating on DataFrames will now also display the column units if available. The `ColumnMetaDataProperty` used in the `CSVSourc` processor got support for specifying units. 
+
+## 2021-12-09 Volume Units
+Volumes now have units and names for both axes and data. The data unit and name is in `volume.dataMapper_.valueAxis.name` and  `volume.dataMapper_.valueAxis.unit` and the axes unit and names are in `volume.axis[i].name` and `volume.axis[i].unit`. 
```
suggestion:
unit and name _are stored_ in

martinfalk commented on CHANGELOG.md:
```diff
@@ -1,6 +1,15 @@
+## 2021-12-09 Volume Units
+Volumes now have units and names for both axes and data. The data unit and name is in `volume.dataMapper_.valueAxis.name` and  `volume.dataMapper_.valueAxis.unit` and the axes unit and names are in `volume.axis[i].name` and `volume.axis[i].unit`. 
+Inviwo now uses the Units library from LLNL (https://github.com/LLNL/units). The library handles parsing units from string `v = units::unit_from_string("m/s")` it can also handle unit calculations, i.e. `v * v`. 
+For printing units we have added fmt support. see `inviwo/include/inviwo/core/datastructures/unitsystem.h` for details about the formatting. 
```
typo: "See ..."

martinfalk commented on CHANGELOG.md:
```diff
@@ -1,6 +1,15 @@
+For printing units we have added fmt support. see `inviwo/include/inviwo/core/datastructures/unitsystem.h` for details about the formatting. 
+Some of the processor now also handle showing units, like the `VolumeAxis`, and `ColorScaleLegend`. The transfer function editor will also show units if available. 
+The Dat and Ivf volume file readers also got support for setting axis/data names and units. 
+# 2021-11-23 CompositeProperty collapse & expand
```
empty line above heading?

martinfalk commented on modules/webbrowser/tests/regression/web-property-sync/script.py:
```diff
@@ -1,16 +1,28 @@
+    inviwopy.logInfo(f"Waiting for webpage {isLoading.path}: {isLoading.value}"
+                     f" time: {time.perf_counter() - start}")
+
+# Trigger an extra network evaluation to try and make sure the web suff is done
```
typo "stuff"

martinfalk reviewed (dismissed):
minor comments

petersteneteg committed:
Webbrowser: review fix
